src/Text.mo

```diff
@@ -26,30 +45,521 @@ module {
+          }
+        }
+      }
+    }
+  };
+
+  /// Returns `true` if `t` starts with a prefix matching [pattern](#type.Pattern) `p`, otherwise returns `false`.
+  public func startsWith(t : Text, p : Pattern) : Bool {
+    var cs = t.chars();
+    let match = matchOfPattern p;
+    switch (match(cs)) {
+      case (#success) true;
+      case _ false;
+    }
+  };
+
+  /// Returns `true` if `t` ends with a suffix matching [pattern](#type.Pattern) `p`, otherwise returns `false`.
+  public func endsWith(t : Text, p : Pattern) : Bool {
+    let s2 = sizeOfPattern p;
+    if (s2 == 0) return true;
+    let s1 = t.size();
+    if (s2 > s1) return false;
+    let match = matchOfPattern p;
+    var cs1 = t.chars();
+    var diff = s1 - s2;
+    while (diff > 0)  {
+      ignore cs1.next();
+      diff -= 1;
+    };
+    switch (match(cs1)) {
+      case (#success) true;
+      case _ false;
+    }
+  };
+
+  /// Returns `t` with all matches of [pattern](#type.Pattern) `p` replaced by text `r`.
+  public func replace(t : Text, p : Pattern, r : Text) : Text {
+    let match = matchOfPattern p;
+    let size = sizeOfPattern p;
+    let cs = CharBuffer(t.chars());
+    var res = "";
+    label l
+    loop {
+      switch (match(cs)) {
+        case (#success) {
+          res #= r;
+          if (size > 0) {
+            continue l;
+          }
+        };
+        case (#empty cs1) {
+          for (c1 in cs1) {
+            res #= fromChar c1;
+          };
+          break l;
+        };
+        case (#fail (cs1, c)) {
+          cs.pushBack(cs1, c);
+        }
+      };
+      switch (cs.next()) {
+        case null {
+          break l;
+        };
+        case (?c1) {
+         res #= fromChar c1;
+        }; // continue
+      }
+    };
+    return res;
+  };
+
+
+
+  /// Returns the optioned suffix of `t` obtained by eliding exactly one leading match of [pattern](#type.Pattern) `p`, otherwise `null`.
+  public func stripStart(t : Text, p : Pattern) : ?Text {
+    let s = sizeOfPattern p;
+    if (s == 0) return ?t;
+    var cs = t.chars();
+    let match = matchOfPattern p;
+    switch (match(cs)) {
+      case (#success) return ?fromIter(cs);
+      case _ return null;
+    }
+  };
+
+  /// Returns the optioned prefix of `t` obtained by eliding exactly one trailing match of [pattern](#type.Pattern) `p`, otherwise `null`.
+  public func stripEnd(t : Text, p : Pattern) : ?Text {
+    let s2 = sizeOfPattern p;
+    if (s2 == 0) return ?t;
+    let s1 = t.size();
+    if (s2 > s1) return null;
+    let match = matchOfPattern p;
+    var cs1 = t.chars();
+    var diff = s1 - s2;
+    while (diff > 0)  {
+      ignore cs1.next();
+      diff -= 1;
+    };
+    switch (match(cs1)) {
+      case (#success) return ?extract(t, 0, ? (s1 - s2));
+      case _ return null;
+    }
+  };
+
+  /// Returns the suffix of `t` obtained by eliding all leading matches of [pattern](#type.Pattern) `p`.
+  public func trimStartMatches(t : Text, p : Pattern) : Text {
```

Comment:
`trimStartPattern` or trimStartPattern?

My intention was to add trimStart and trimEnd for whitespace later (needs Char.isWhiteSpace in next-moc).

Reply:
I'd rather call the other `trimEndWhitespace` then. Either way, the difference between "strip" and "trim" isn't very obvious, at least to somebody with my English skills. Instead of different verbs, perhaps use some suffix to distinguish?

Reply:
stripManyStart/End? and leave trimStart, trimEnd for whitespace? I think trim is pretty standard.

Reply:
We could just add a #whitespace pattern as shorthand for #predicate Char.isWhiteSpace

Reply:
Do we also want `Trim` for `Trim` both?